File: Medium/11-container-w-most-water.py
# Runtime: 961 ms, faster than 53.47% of Python3 online submissions for Container With Most Water.
# Memory Usage: 27.5 MB, less than 63.22% of Python3 online submissions for Container With Most Water.

class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        maxArea = 0
        
        while left <= right:
            maxArea = max(maxArea, min(height[left], height[right]) * (right - left))
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
                
        return maxArea

File: Medium/11-containerwmostwater.cpp
#include <vector>

class Solution {
public:
    int maxArea(std::vector<int>& height) {
        int lo = 0;
        int hi = height.size() - 1;
        int maxWater = 0;
        
        while (lo != hi) {
            int smaller = std::min(height[lo], height[hi]);
            maxWater = std::max(
                smaller * (hi - lo),
                maxWater
            );
                
            if (height[lo] < height[hi]) {
                lo++;
            } else {
                hi--;
            }
        }
        
        return maxWater;
        
    }
};